### ISSUE TITLE
Create an Apex method to process a packet

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#5588]
**[Mingle Card: 5951](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/5951)**
The Managed Package code has been built into MP 236 and installed on the selenium sandbox s1.

A full selenium suite is running with the Mobile PR on FXQA1 against the new Managed Package.

Back Office PR
--------------

[https://github.com/Liquidframeworks/alpine-package/pull/787](https://github.com/Liquidframeworks/alpine-package/pull/787)
--------------------------------------------------------------------------------------------------------------------------

```
Beta package 
``````
04t1Y000000tFKzQAM \[04t1Y000000tFKpQAM\]
```

Narrative
---------

Provide REST API access to a method that will process an upload packet consisting of many events.

The first approach is to have Mobile create the audit log entry with the upload events in the attachment, then this code will process those events.

The second approach is to upload the events directly to this REST API method.

**With either approach, the Apex code is fully responsible for processing the events and marking the packet with success or failure.**

Acceptance Criteria
-------------------

Create an Apex method, presumably a REST Endpoint, that can be called by Mobile.

The call should be asynchronous, Mobile should be able to initiate the function and then disconnect.

The method should accept an audit log id as an input parameter (assuming we are using the first approach)

This method should open the attachment and process each event, one at a time, in order.

If a single event has an error, then it should be written to the Sync Event object.

The processor should continue to process events even if a preceding event failed.

All failed events should be written to the Sync Event object.

When all of the events in the audit log attachment have been processed, whether successful or failed, the audit log record should be marked as Processed. (Type\_\_c = ‘Bulk Upload Processed’)

At this point, every event should have either been successfully run, or should exist in the Sync Event object.

This method will need to support packets with 1,000 events.  

_Analysis_
----------

**Known limits:**

1.  6 MB of heap, which means we must be mindful of report attachments when we are sending them up.  Possible solutions could involve sending them separately or even after the packet is processed since latter events can have no dependency on the existence of attachments.
2.  There is a 100 request limit and 150 write limit for storage operations in a piece of apex code.  Since all updates and deletes require a select to get the identifier before the write can happen, and we want to keep the initial code simple, we should assume all events will cost one read operation.  That leaves us with a limit of 100 events.  We can work around this by having the mobile client break the packet into 100 event chunks.

The packets could be split into batches of 100 and possible all attached to the same audit log if we like.  It would be ideal if the audit log entries “status” could be changed to complete once the packet handler is done processing all of the events.

Note – rluo

we need to consider sfdc governor limits 

[https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex\_gov\_limits.htm](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_gov_limits.htm)

Each apex transaction only allows 100 soql queries and 150 insert/update/delete DML. For each failed insert/update/delete, we need to insert AuditLog or SyncEvent and we could batch the AuditLog/SyncEvent insertion.

*   The packets could not be more than 100 events

Related Stories
---------------

Tasks
-----

{{ table query: SELECT Number, Name, Owner, 'Task Status' WHERE Type = Task and Story = THIS CARD ORDER BY Status }}

Defects
-------

{{ table query: SELECT Number, Name, Owner, 'Status' WHERE Type = Defect and 'Related Story' = THIS CARD }}

Test Areas
----------

_Developers: During code review, delete the test areas that do not apply to this card, leave the test areas that should be covered by QA._

*   Audit logs
*   Incremental Sync
*   Mixpanel
*   Reports
*   Rich Text field
*   Status and Status Workflow

Test Plan
---------

You may need to add a custom setting called SyncPacketSize\_\_c.  By default this value is 60, you may need to set it as low as 20.

1.  Test the post process with 300 events and get benchmark time between background sync and this new sync.
2.  Test the sync with a very large report.
3.  login to mobile, go offline, create data :
    1.  create new job, add ticket, add items, bulk edit items
    2.  create new ticket on a existing job that was synced to backoffice
    3.  Make sure you make more than 80 new items or updates. 
    4.  go to backoffice and delete job used in step 3.2. 
    5.  in mobile go online to initiate background sync 
    6.  once sync completes
    7.  expected: valid data synced up but the ticket and items with the deleted job should be in the sync\_events table
    8.  there are audit logs for each completed sync packet; number of audit logs depends on number of events created in mobile
        1.  (one audit log for each X events; any action in mobile counts as event. The exact count depends on your SyncPacketSize custom setting. The default is 60 events per audit log packet.
4.  Set SyncPacketSize to a large number (200 for example)  
    login to mobile, go offline, create data to generate large number of events  
    (add a few tickets, each should have 40 items, bulk edit 2-3 fields on ticket items in each ticket to generate at least 200 events)  
    go online to initiate background sync  
    expected: data not synced up bc packet size larger then SF limits  
                    data still present on mobile device  
                    there are audit logs that state that sync was not complete. The audit log has the data packet attached.  
                    there are lines in LogEntries to indicate that the process failed.

  
**Test Result: Pass**

```
Beta package 04t1Y000000tFKzQAM 
```

**Mobile PR: 2863**

Mobile User: lanrefield@seven.com.s3   /  lastpass

Back Office user: qa@seven.com.s3  / lastpass

Hardware: Dell XPS PC, IPAD Pro.

OS: Windows 10 & IOS 11

Browser: Chrome & Safari 

   
 

  
 

**Test Case: Updated **Background Sync - Use Background Sync for Upload Only and Sync 

[https://lfw.testrail.io/index.php?/cases/view/13478](https://lfw.testrail.io/index.php?/cases/view/13478)

**Test Notes:** TheNew sync requires the back office admin to set SyncPacketSize\_\_c to 40 or less in custom settings. 

See below for performance test results.

*   **Old post process with 300 ticket items with old background sync**  [https://hunterpeaksandbox.herokuapp.com](https://hunterpeaksandbox.herokuapp.com)
    -------------------------------------------------------------------------------------------------------------------------------------------------------
    

**Windows 10 Chrome – ****Sync was 5 mins 16 secs **

!oldsync2.JPG!{height: 340px; width: 950px;}

**Two Large reports with 150 items each and large image attachment on the old sync  –  ****Sync was 2.52 seconds**

!Capturea-fb607f.JPG!{height: 274px; width: 950px;}

**IPAD IOS 11 Safari   –  ****Sync was 3 mins 52 secs**

!00723BD5-D238-4917-98BF-D6C6DDDC2B0B.png!{height: 675px; width: 700px;}+

*   **New Post process with 300 events with new background sync on PR 2863 **[https://sandbox.fieldfx.com/](https://sandbox.fieldfx.com/)
    -------------------------------------------------------------------------------------------------------------------------------------
    

**Windows 10 Chrome  –  ****Sync was 1 min 31 secs. A ****Secondary test was 1 Min 14 secs. ** 

!Newsync.png!{height: 362px; width: 1000px;}

!Newsync20.JPG!{width: 950px;}

**Two Large reports with 150 items each and large image attachment on the new sync  –  Sync was 1.38 seconds on windows chrome and 1.13 seconds on the IPAD.**

!Captureax.JPG!{height: 275px; width: 900px;}

**IPAD IOS 11 Safari – ****Sync was 58 secs.  **

!captures.JPG!{height: 751px; width: 950px;}

**Back office data set**

!Capturex.JPG!{height: 452px; width: 800px;}

*   **Step 3 Scenario **

** Valid data synced up. 80 + Events **

 !Capture11.png!{height: 352px; width: 750px;}

**Audit logs for each completed sync packet**

!Capture-2939a5.JPG!{height: 147px; width: 900px;}

**Sync Event Viewer for deleted job**

!Capture-f03659.JPG!{height: 143px; width: 850px;}

!Capture2-321108.JPG!{height: 129px; width: 900px;}

*   **Step 4 Set SyncPacketSize to a large number (200 for example)**

**Data not synced up bc packet size larger then SF limits**

!Capture-146f77.JPG!{height: 269px; width: 750px;}

**Audit logs state that sync was not complete and audit log has the data packet attached.**

!Capture2-bc8927.JPG!{height: 68px; width: 750px;}

!Capture11-bad2b9.JPG!{height: 497px; width: 700px;}

!Capture22-fa2908.JPG!{height: 268px; width: 750px;}

**Comment by githubbob42** [comment]
Lanre : 09/20/2018 13:55 CDT
Sync fails CODE: APEX\_ERROR SALESFORCE: System.LimitException: Too many SOQL queries: 101 ()

**Comment by githubbob42** [comment]
Lanre : 10/01/2018 15:08 CDT
Needs Selenium

**Comment by githubbob42** [comment]
**Pull Request #2863**
